### PR TITLE
Direct support for compiled models

### PR DESCRIPTION
**Pull request by lantiga**
## What does this PR do?

This PR introduces dynamo/inductor support to Lightning. A prior design proposed the use of a callback.

However, since compilation is going to be first-class in PyTorch, through `torch.compile`, here we add direct support for models compiled through `torch.compile`, so there's zero cognitive overhead for people who get into compiling PyTorch models. This is what it looks like, from regular modules:

```python
model = BoringModel()

trainer = Trainer()
trainer.fit(model)
```

to compiled ones:

```python
model = BoringModel()
model = torch.compile(model)

trainer = Trainer()
trainer.fit(model)
```

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/core/module.py [hunk 1]
```diff
@@ -1950,6 +1951,70 @@ def _register_sharded_tensor_state_dict_hooks_if_available(self) -> None:
...
+            "compiler": "dynamo",
+            "dynamo_ctx": model.dynamo_ctx,
+            "original_forward": orig_module.forward,
+        }
+
+        orig_module.forward = model.dynamo_ctx(orig_module.forward)  # type: ignore[assignment]
+        return orig_module
+
+    @classmethod
+    def to_uncompiled(cls, model: Union["pl.LightningModule", "torch._dynamo.OptimizedModule"]) -> "pl.LightningModule":
```
Comment (tchaton) [inline src/pytorch_lightning/core/module.py:1989]: This isn't used yet right ?

File: src/pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -963,6 +988,18 @@ def _restore_modules_and_callbacks(self, checkpoint_path: Optional[_PATH] = None
     def _run(
         self, model: "pl.LightningModule", ckpt_path: Optional[str] = None
     ) -> Optional[Union[_EVALUATE_OUTPUT, _PREDICT_OUTPUT]]:
+        if model._compiler_ctx is not None:
```
Comment (tchaton) [inline src/pytorch_lightning/trainer/trainer.py:991]: I think it would be cleaner for the strategy to implement a validation mechanism directly and for DDPStrategy / DDPFullyShardedNativeStrategy to uncompile the model and wrap their own forward model after wrapping.

File: src/pytorch_lightning/core/module.py (hunk 1 above)
Comment (lantiga) [reply src/pytorch_lightning/core/module.py:1989]: Only in tests for now

**Review by tchaton — APPROVED** [review]
LGTM !

File: src/pytorch_lightning/core/module.py [hunk 3]
```diff
@@ -1950,6 +1951,70 @@ def _register_sharded_tensor_state_dict_hooks_if_available(self) -> None:
                 weakref.proxy(self), pre_load_state_dict_hook, True  # type: ignore[arg-type]
             )
 
+    @classmethod
+    def from_compiled(cls, model: Union["pl.LightningModule", "torch._dynamo.OptimizedModule"]) -> "pl.LightningModule":
```
Comment (ethanwharris) [inline src/pytorch_lightning/core/module.py:1955]: Typed as LightningModule or OptimizedModule but then raises an error for anything that isn't an OptimizedModule:
```suggestion
    def from_compiled(cls, model: "torch._dynamo.OptimizedModule") -> "pl.LightningModule":
```

File: src/pytorch_lightning/core/module.py [hunk 4]
```diff
@@ -1950,6 +1951,70 @@ def _register_sharded_tensor_state_dict_hooks_if_available(self) -> None:
...
+                "`model` is expected to be a compiled LightingModule. " f"Found a compiled {type(orig_module)} instead"
+            )
+
+        orig_module._compiler_ctx = {
+            "compiler": "dynamo",
+            "dynamo_ctx": model.dynamo_ctx,
+            "original_forward": orig_module.forward,
+        }
+
+        orig_module.forward = model.dynamo_ctx(orig_module.forward)  # type: ignore[assignment]
```
Comment (tchaton) [inline src/pytorch_lightning/core/module.py:1985]: Let's wrap `training_step`, some users might not use forward in their model.

**Review by ethanwharris — APPROVED** [review]
LGTM, very minor comment

**Review by justusschock — APPROVED** [review]
the code here looks good to me,

a general question: does ddp(compiled model) work? should we in that case call compile(ddp(orig model))? to allow for possibly better optimizations in distributed settings? what about all the other strategies?

File: src/pytorch_lightning/core/module.py (hunk 3 above)
Comment (lantiga) [reply src/pytorch_lightning/core/module.py:1955]: great catch!

**Comment by lantiga** [comment]
> the code here looks good to me,
> 
> a general question: does ddp(compiled model) work? should we in that case call compile(ddp(orig model))? to allow for possibly better optimizations in distributed settings? what about all the other strategies?

Excellent questions, probably compile(ddp(orig model)) is the way to go for distributed.
For now we want to unblock this this way and understand the best practices around distributed over the next period.

We only allow DDP and FSDP for now, which are the ones that are supported so far.

File: src/pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (lantiga) [reply src/pytorch_lightning/trainer/trainer.py:991]: Agreed, this will be the way forward in the future, as soon as we understand more about dynamo and distributed.

File: src/pytorch_lightning/core/module.py (hunk 4 above)
Comment (lantiga) [reply src/pytorch_lightning/core/module.py:1985]: of course, that makes sense

Comment (lantiga) [reply src/pytorch_lightning/core/module.py:1985]: I'm kind of hesitant because training_step contains things that dynamo doesn't seem to be tested on right now. I propose we merge this now and work on wrapping training_step later as the technology matures

File: src/pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (carmocca) [reply src/pytorch_lightning/trainer/trainer.py:991]: @lantiga I agree with @tchaton. This could be implemented as a [plugin](https://github.com/Lightning-AI/lightning/tree/master/src/pytorch_lightning/plugins) that is specifically enabled for some of the strategies, Which would contain all this patching and unpatching logic.

If you agree, I would like to give it a shot
